src/main/java/io/r2dbc/postgresql/client/ExtendedQueryMessageFlow.java

```diff
@@ -69,15 +76,55 @@ private ExtendedQueryMessageFlow() {
      * @return the messages received in response to the exchange
      * @throws IllegalArgumentException if {@code bindings}, {@code client}, {@code portalNameSupplier}, or {@code statementName} is {@code null}
      */
-    public static Flux<BackendMessage> execute(Publisher<Binding> bindings, Client client, PortalNameSupplier portalNameSupplier, String statementName, String query, boolean forceBinary) {
-        Assert.requireNonNull(bindings, "bindings must not be null");
+    public static Flux<BackendMessage> execute(Binding binding, Client client, PortalNameSupplier portalNameSupplier, String statementName, String query, boolean forceBinary,
+                                               int fetchSize) {
+        Assert.requireNonNull(binding, "binding must not be null");
         Assert.requireNonNull(client, "client must not be null");
         Assert.requireNonNull(portalNameSupplier, "portalNameSupplier must not be null");
         Assert.requireNonNull(statementName, "statementName must not be null");
-
-        return client.exchange(Flux.from(bindings)
-            .flatMap(binding -> toBindFlow(binding, portalNameSupplier, statementName, query, forceBinary))
-            .concatWith(Mono.just(Sync.INSTANCE)));
+        if (fetchSize == NO_LIMIT) {
+            return Flux.defer(() -> {
+                String portal = portalNameSupplier.get();
+                Flux<FrontendMessage> bindFlow = toBindFlow(binding, portal, statementName, query, forceBinary, fetchSize).concatWithValues(new Close(portal, PORTAL))
+                    .concatWithValues(Sync.INSTANCE);
+                return client.exchange(bindFlow)
+                    .as(Operators::discardOnCancel)
+                    .doOnDiscard(ReferenceCounted.class, ReferenceCountUtil::release);
+            });
+        }
+        return Flux.defer(() -> {
+            String portal = portalNameSupplier.get();
+            Flux<FrontendMessage> bindFlow = toBindFlow(binding, portal, statementName, query, forceBinary, fetchSize).concatWithValues(Flush.INSTANCE);
+            DirectProcessor<FrontendMessage> requestsProcessor = DirectProcessor.create();
```

Comment:
How about using just `EmitterProcessor` without an additional sink? We can avoid another queue creation with `EmitterProcessor`.

Reply:
`EmitterProcessor` has some queue inside, `DirectProcessor` has no. We should never have situation when `DirectProcessor` has no subscribers.